agency/agency_connection.go

```diff
@@ -0,0 +1,295 @@
+}
+
+// isArangoError checks if the given error is (or is caused by) an ArangoError.
+// If so it returned the Code and true, otherwise it returns 0, false.
+func isArangoError(err error) (int, bool) {
+	if aerr, ok := driver.Cause(err).(driver.ArangoError); ok {
+		return aerr.Code, true
+	}
+	return 0, false
+}
+
+// Unmarshal unmarshals the given raw object into the given result interface.
+func (c *agencyConnection) Unmarshal(data driver.RawObject, result interface{}) error {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	if len(c.connections) == 0 {
+		return driver.WithStack(fmt.Errorf("no connections"))
+	}
+	if err := c.connections[0].Unmarshal(data, result); err != nil {
+		return driver.WithStack(err)
+	}
+	return nil
+}
+
+// Endpoints returns the endpoints used by this connection.
+func (c *agencyConnection) Endpoints() []string {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	var result []string
+	for _, x := range c.connections {
+		result = append(result, x.Endpoints()...)
+	}
+	return result
+}
+
+// UpdateEndpoints reconfigures the connection to use the given endpoints.
+func (c *agencyConnection) UpdateEndpoints(endpoints []string) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	newConnections := make([]driver.Connection, len(endpoints))
+	for i, ep := range endpoints {
+		config := c.config
+		config.Endpoints = []string{ep}
+		config.FailOnRedirect = true
+		httpConn, err := http.NewConnection(config)
+		if err != nil {
+			return driver.WithStack(err)
+		}
+		if c.auth != nil {
+			httpConn, err = httpConn.SetAuthentication(c.auth)
+			if err != nil {
+				return driver.WithStack(err)
+			}
+		}
+		newConnections[i] = httpConn
+	}
+	c.connections = newConnections
+	return nil
+}
+
+// Configure the authentication used for this connection.
+func (c *agencyConnection) SetAuthentication(auth driver.Authentication) (driver.Connection, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	newConnections := make([]driver.Connection, len(c.connections))
+	for i, x := range c.connections {
+		xAuth, err := x.SetAuthentication(auth)
+		if err != nil {
+			return nil, driver.WithStack(err)
+		}
+		newConnections[i] = xAuth
+	}
+	c.connections = newConnections
+	c.auth = auth
+	return c, nil
+}
+
+// Protocols returns all protocols used by this connection.
+func (c *agencyConnection) Protocols() driver.ProtocolSet {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	result := driver.ProtocolSet{}
+	for _, x := range c.connections {
+		for _, p := range x.Protocols() {
+			if !result.Contains(p) {
+				result = append(result, p)
+			}
+		}
+	}
+	return result
+}
+
+// agencyConnectionFailureBackoff returns a backoff delay for cases where all
+// agents responded with a non-fatal error.
+func agencyConnectionFailureBackoff(lastDelay time.Duration) time.Duration {
+	return increaseDelay(lastDelay, 1.5, time.Millisecond, time.Second*2)
```

Comment:
I fear that a maximum delay of 2 seconds will not be enough in cases in which the machine is under a heavy load. I suggest to increase the maximum considerably to maybe 15 or 30 seconds.

Reply:
I don't understand why. The result of this function is used as a delay period between failing requests. It is not the timeout for an agency request.

Reply:
OK, then I misunderstood this, sorry. Please ignore.